Microsoft.Toolkit.Mvvm/ComponentModel/ObservableValidator.cs

```diff
@@ -0,0 +1,159 @@
+            if (string.IsNullOrEmpty(propertyName))
+            {
+                return this.GetAllErrors();
+            }
+
+            // Property-level errors, if any
+            if (this.errors.TryGetValue(propertyName!, out List<ValidationResult> errors))
+            {
+                return errors;
+            }
+
+            // The INotifyDataErrorInfo.GetErrors method doesn't specify exactly what to
+            // return when the input property name is invalid, but given that the return
+            // type is marked as a non-nullable reference type, here we're returning an
+            // empty array to respect the contract. This also matches the behavior of
+            // this method whenever errors for a valid properties are retrieved.
+            return Array.Empty<ValidationResult>();
+        }
+
+        /// <summary>
+        /// Implements the logic for entity-level errors gathering for <see cref="GetErrors"/>.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerable"/> instance with all the errors in <see cref="errors"/>.</returns>
+        [Pure]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private IEnumerable GetAllErrors()
+        {
+            return this.errors.Values.SelectMany(errors => errors);
+        }
+
+        /// <summary>
+        /// Validates a property with a specified name and a given input value.
+        /// </summary>
+        /// <param name="value">The value to test for the specified property.</param>
+        /// <param name="propertyName">The name of the property to validate.</param>
+        /// <returns>Whether or not the validation was successful.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when <paramref name="propertyName"/> is <see langword="null"/>.</exception>
+        protected bool ValidateProperty(object? value, [CallerMemberName] string? propertyName = null)
+        {
+            if (propertyName is null)
+            {
+                ThrowArgumentNullExceptionForNullPropertyName();
+            }
+
+            bool errorsChanged = false;
+
+            // Clear the errors for the specified property, if any
+            if (this.errors.TryGetValue(propertyName!, out List<ValidationResult>? propertyErrors) &&
+                propertyErrors.Count > 0)
+            {
+                propertyErrors.Clear();
+
+                errorsChanged = true;
+            }
+
+            List<ValidationResult> results = new List<ValidationResult>();
+
+            // Validate the property
+            bool isValid = Validator.TryValidateProperty(
+                value,
+                new ValidationContext(this, null, null) { MemberName = propertyName },
+                results);
+
+            // Update the state and/or the errors for the property
+            if (isValid)
+            {
+                if (errorsChanged)
+                {
+                    this.totalErrors--;
```

Comment:
I'm not exactly sure what this count value is for. Can we not use the errors dictionary as the source of truth?

Reply:
I've added an XML comment for that here:

https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/be109be7a7c1b03c9fac2fa51e6b1f4c398f5a48/Microsoft.Toolkit.Mvvm/ComponentModel/ObservableValidator.cs#L27-L32

My idea here was - with how the code is currently setup, the dictionary can contain entries for valid properties, just with the list of errors being empty (so that we can reuse the same `List<ValidationResult>` instances when possible. So in order to check whether we have errors, we'd have to traverse the entire dictionary and check whether all values are empty lists (or if no entries are there at all). This would have a cast of `O(N)`, as you'd have to traverse the entire dictionary once to check all properties, and the dictionary could potentially contain all the existing properties for a given object. So the point of this additional field, which is kept in sync with the dictionary, is to provide a fast `O(1)` accessor to quickly check whether or not there are any errors in the current instance. Let me know what you think! 😊

Reply:
I appreciate the attention to perf, but how often do we expect errors to transition between valid/invalid? 

My gut is that we won't really be allocating that often to warrant this extra bit of complexity and size trade off. Computers are much faster at allocating/deallocating memory than humans are at typing :)

Reply:
Yeah that's a good point, I've removed the field in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3429/commits/e7a482458b9255d594dc4302dd1732c4e22bacc8 😊